From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: violetc <58360096+s-yh-china@users.noreply.github.com>
Date: Sun, 9 Jul 2023 16:51:47 +0800
Subject: [PATCH] Raider die skip self raid check


diff --git a/net/minecraft/world/entity/raid/Raider.java b/net/minecraft/world/entity/raid/Raider.java
index c06b589e669b055a26f662df60070d5908256220..11555c7770b1dbe3d4ff55470df513c4039ff124 100644
--- a/net/minecraft/world/entity/raid/Raider.java
+++ b/net/minecraft/world/entity/raid/Raider.java
@@ -153,7 +153,7 @@ public abstract class Raider extends PatrollingMonster {
     }
 
     public boolean hasRaid() {
-        return this.level() instanceof ServerLevel serverLevel && (this.getCurrentRaid() != null || serverLevel.getRaidAt(this.blockPosition()) != null);
+        return !org.dreeam.leaf.config.modules.gameplay.RaidConfig.skipSelfRaidCheck && (this.level() instanceof ServerLevel serverLevel && (this.getCurrentRaid() != null || serverLevel.getRaidAt(this.blockPosition()) != null));  // Leaves - skip self raid check
     }
 
     public boolean hasActiveRaid() {
